[sparse] Add i8i8->i32 support for cuSPARSELt

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110499

Summary:

With the release of cuSPARSELt v0.5.0, we now have support for
int8 int8 -> int32 matmul.

This PR adds support for this via out_dtype.

Test Plan:
```
python test/test_sparse_semi_structured.py -k int32
```

Reviewers:

Subscribers:

Tasks:

Tags:

jcaip committed:
[sparse] Add i8i8->i32 support for cuSPARSELt

Summary:

With the release of cuSPARSELt v0.5.0, we now have support for
int8 int8 -> int32 matmul.

This PR adds support for this via out_dtype.

Test Plan:
```
python test/test_sparse_semi_structured.py -k int32
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add i8i8->i32 support for cuSPARSELt"


Summary:

With the release of cuSPARSELt v0.5.0, we now have support for
int8 int8 -> int32 matmul.

This PR adds support for this via out_dtype.

Test Plan:
```
python test/test_sparse_semi_structured.py -k int32
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse_semi_structured.py:
```diff
@@ -233,6 +233,21 @@ def test_cslt_sparse_mm_int8_in_fp16_out(self, device):
+
+            B = torch.rand((128, 128), device=A_sparse.device).to(torch.int8)
+
+            dense_result = torch.mm(A.cpu(), B.t().cpu()).to(device, dtype=torch.int32)
```
torch.mm (ironically) produces int8 results as well when given int8 inputs. I'd suggest to use int64 as a reference.

cpuhrsch reviewed (changes requested):
Requesting changes just to make sure the comment doesn't get lost

jcaip committed:
Update on "[sparse] Add i8i8->i32 support for cuSPARSELt"


Summary:

With the release of cuSPARSELt v0.5.0, we now have support for
int8 int8 -> int32 matmul.

This PR adds support for this via out_dtype.

Test Plan:
```
python test/test_sparse_semi_structured.py -k int32
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse_semi_structured.py:
```diff
@@ -233,6 +233,21 @@ def test_cslt_sparse_mm_int8_in_fp16_out(self, device):
+
+            B = torch.rand((128, 128), device=A_sparse.device).to(torch.int8)
+
+            dense_result = torch.mm(A.cpu(), B.t().cpu()).to(device, dtype=torch.int64)
```
I still don't think this fixes it.

You'd have to do 
```
dense_result = torch.mm(A.to(torch.int64).cpu(), B.to(torch.int64).cpu()).to(device)
assert dense_result.dtype = torch.int64
```

cpuhrsch replied:
I don't see how else we can check that indeed cuSPARSElt is now returning i32 instead of i8

jcaip committed:
Update on "[sparse] Add i8i8->i32 support for cuSPARSELt"


Summary:

With the release of cuSPARSELt v0.5.0, we now have support for
int8 int8 -> int32 matmul.

This PR adds support for this via out_dtype.

Test Plan:
```
python test/test_sparse_semi_structured.py -k int32
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
@pytorchbot merge

jcaip commented:
@pytorchbot revert this as cslt v0.5.0 requires a newer linker and we will be using v0.4.0 as the base version

jcaip commented:
@pytorchbot revert -m cslt v0.5.0 requires a newer linker and we will be using v0.4.0 as the base version -c weird

jcaip commented:
@pytorchbot revert -m "cslt v0.5.0 requires a newer linker and we will be using v0.4.0 as the base version" -c weird
